docs/proposals/20200304-autoscaler-scale-ux.md

```diff
@@ -0,0 +1,117 @@
+---
+title: Cluster Autoscaler Scale from Zero UX
+authors:
+  - "@michaelgugino"
+reviewers:
+  - "@detiber"
+creation-date: 2020-03-04
+creation-date: 2020-03-04
+status:
+---
+
+# Cluster Autoscaler Scale from Zero UX
+
+## Summary
+Cluster Autoscaler supports scale from zero for cloud-native providers. This is done via compiling CPU, GPU, and RAM info into the cluster autoscaler at build time for each supported cloud.
+
+We are attempting to add cluster-api support to the cluster-autoscaler to scale machinesets, machinedeployments, and/or other groupings of machines as appropriate. We are trying to keep an abstraction which is isolated from particular cloud implementations. We need a way to inform the cluster-autoscaler of the value of machine resources so it can make proper scheduling decisions when there are not currently any machines in a particular scale set (machineset/machinedeployment, etc).
+
+We will need input from the autoscaler team, as well as input from api reviewers to ensure we obtain the best solution.
+
+These ideas require a particular (possibly new) cluster-api cloud provider controller
+which can inspect machinesets and annotate them as appropriate.  This could be
+done dynamically via billing API (not recommended) or have values compiled in
+statically.
+
+We could also attempt to build this logic into the machineset/machinedeployment
+controller, but this would probably not be very scalable and would require
+a well defined set of fields on the provider CRD.
+
+## Motivation
+
+### Goals
+
+### Non-Goals
```

Comment:
Perhaps this is only because I haven't been following prior conversations about this (sorry!), but I think I would be better able to understand the proposal if the motivation section included some goals and non-goals.

For example, is it a goal to minimize (or not have any?) impact on any cluster API provider which does not support autoscale from zero? Is it a goal to minimize impact/knowledge burden for cluster API users who are not using this feature?

Reply:
The goal is to support scale from zero in an automated way.  We'll be relying on providers to provide the attributes (CPU, RAM, GPU) for their respective instance types.  We also want to build a mechanism to support scale from zero on cloud providers that don't have common billing (such as an openstack cloud or some other abstraction) by providing a field (either spec or annotation on some object) that allows an administrator to define the attributes themselves.